File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root,bool &ans){
        if(root == NULL)return 0;

        int left = maxDepth(root->left,ans);
        int right = maxDepth(root->right,ans);

        if(abs(left-right) > 1)ans = false;

        return 1+max(left,right);

    }
    bool isBalanced(TreeNode* root) {
        if(root == NULL)return true;

        bool ans = true;
        maxDepth(root,ans);
        
        return ans;
    }
};

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 // Inorder to get the direct height we need to capture the leaf node at the base case
 // if(root == null)return 0; and then if(root.left == null && root.right == null)return 0;
 // Otherwise calculate maxDepth and reduce by one to find the no of edges
class Solution {
    boolean isTreeBalanced = true;
    public int maxDepth(TreeNode root){
        if(root == null)return 0;

        int maxDepthLeftSubtree = maxDepth(root.left);
        int maxDepthRightSubtree = maxDepth(root.right);

        if(Math.abs(maxDepthLeftSubtree-maxDepthRightSubtree) > 1){
            isTreeBalanced = false;
        }

        return Math.max(maxDepthRightSubtree,maxDepthLeftSubtree) + 1;
    }
    public boolean isBalanced(TreeNode root) {
        // no use of return value just call
        maxDepth(root);

        return isTreeBalanced;
    }
}